File: api/app/services/source_service.py
import uuid

from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.source import Source
from app.schemas.source import SourceCreate, SourceUpdate


async def create_source(session: AsyncSession, data: SourceCreate) -> Source:
    dump = data.model_dump()
    dump["rss_url"] = str(data.rss_url)
    if data.site_url:
        dump["site_url"] = str(data.site_url)
    source = Source(**dump)
    session.add(source)
    await session.flush()
    return source


async def get_sources(
    session: AsyncSession,
    page: int = 1,
    per_page: int = 20,
    active_only: bool = False,
) -> tuple[list[Source], int]:
    query = select(Source)
    if active_only:
        query = query.where(Source.is_active.is_(True))

    count_result = await session.execute(
        select(func.count()).select_from(query.subquery())
    )
    total = count_result.scalar_one()

    result = await session.execute(
        query.order_by(Source.created_at.desc())
        .offset((page - 1) * per_page)
        .limit(per_page)
    )
    sources = list(result.scalars().all())
    return sources, total


async def get_source_by_id(
    session: AsyncSession, source_id: uuid.UUID
) -> Source | None:
    result = await session.execute(select(Source).where(Source.id == source_id))
    return result.scalar_one_or_none()


async def update_source(
    session: AsyncSession,
    source: Source,
    data: SourceUpdate,
) -> Source:
    update_data = data.model_dump(exclude_unset=True)
    if "rss_url" in update_data and update_data["rss_url"] is not None:
        update_data["rss_url"] = str(update_data["rss_url"])
    if "site_url" in update_data and update_data["site_url"] is not None:
        update_data["site_url"] = str(update_data["site_url"])
    for key, value in update_data.items():
        setattr(source, key, value)
    await session.flush()
    return source


async def deactivate_source(
    session: AsyncSession,
    source: Source,
) -> Source:
    source.is_active = False
    await session.flush()
    return source

File: api/src/services/source-service.ts
import { count, desc, eq } from "drizzle-orm";
import type { DB } from "../database.js";
import { sources } from "../db/schema/index.js";
import type { SourceCreate, SourceUpdate } from "../schemas/source.js";

// Column-aware type for partial updates — NOT NULL columns exclude null
interface SourceUpdateData {
  name?: string;
  rssUrl?: string;
  siteUrl?: string | null;
  category?: string | null;
  isActive?: boolean;
}

export async function createSource(db: DB, data: SourceCreate) {
  const [source] = await db
    .insert(sources)
    .values({
      name: data.name,
      rssUrl: data.rss_url,
      siteUrl: data.site_url ?? null,
      category: data.category ?? null,
    })
    .returning();
  return source;
}

export async function getSources(
  db: DB,
  page = 1,
  perPage = 20,
  activeOnly = false,
): Promise<{ items: (typeof sources.$inferSelect)[]; total: number }> {
  const whereClause = activeOnly ? eq(sources.isActive, true) : undefined;

  const [totalResult] = await db.select({ count: count() }).from(sources).where(whereClause);
  const total = totalResult.count;

  const items = await db
    .select()
    .from(sources)
    .where(whereClause)
    .orderBy(desc(sources.createdAt))
    .offset((page - 1) * perPage)
    .limit(perPage);

  return { items, total };
}

export async function getSourceById(db: DB, sourceId: string) {
  const [source] = await db.select().from(sources).where(eq(sources.id, sourceId));
  return source ?? null;
}

export async function updateSource(db: DB, sourceId: string, data: SourceUpdate) {
  // Build update object from only defined (sent) fields
  // undefined = not sent, null = explicitly null (rejected by schema for required fields)
  const updateData: SourceUpdateData = {};

  // NOT NULL columns: double-guard null (Zod refine already prevents, this narrows the type)
  if (data.name != null) updateData.name = data.name;
  if (data.rss_url != null) updateData.rssUrl = data.rss_url;
  // Nullable columns: null clears the value
  if (data.site_url !== undefined) updateData.siteUrl = data.site_url ?? null;
  if (data.category !== undefined) updateData.category = data.category ?? null;
  // NOT NULL with default
  if (data.is_active != null) updateData.isActive = data.is_active;

  if (Object.keys(updateData).length === 0) {
    // No fields to update, return existing
    return getSourceById(db, sourceId);
  }

  const [updated] = await db
    .update(sources)
    .set(updateData)
    .where(eq(sources.id, sourceId))
    .returning();
  return updated;
}

export async function deactivateSource(db: DB, sourceId: string) {
  const [deactivated] = await db
    .update(sources)
    .set({ isActive: false })
    .where(eq(sources.id, sourceId))
    .returning();
  return deactivated;
}
